## Quasi-constant features

Quasi-constant features are those that show the same value for the great majority of the observations of the dataset. In general, these features provide little, if any, information that allows a machine learning model to discriminate or predict a target. But there can be exceptions. So you should be careful when removing these type of features.

Identifying and removing quasi-constant features, is an easy first step towards feature selection and more interpretable machine learning models.

Here, we will demonstrate how to identify quasi-constant features using a dataset.

To identify quasi-constant features, we can use the VarianceThreshold from Scikit-learn, or we can code it ourselves. If we use the VarianceThreshold, all our variables need to be numerical. If we code it manually however, we can apply the code to both numerical and categorical variables.



In [ ]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split

from sklearn.feature_selection import VarianceThreshold

In [ ]:
# load dataset

# (feel free to write some code to explore the dataset and become
# familiar with it ahead of this demo)

data = pd.read_csv('../dataset_1.csv')
data.shape

(50000, 301)

**Important**

In all feature selection procedures, it is good practice to select the features by examining only the training set. And this is to avoid overfit.

In [ ]:
# separate dataset into train and test

X_train, X_test, y_train, y_test = train_test_split(
    data.drop(labels=['target'], axis=1), # drop the target
    data['target'], # just the target
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((35000, 300), (15000, 300))

## Remove constant features

First, we will remove constant features like we did in the previous lecture. This will allow a better visualisation of the quasi-constant ones.

In [ ]:
# Remove constant features



((35000, 266), (15000, 266))

## Remove quasi-constant features

### Using the VarianceThreshold from sklearn

The VarianceThreshold from sklearn provides a simple baseline approach to feature selection. It removes all features which variance doesn’t meet a certain threshold. By default, it removes all zero-variance features, as we did in the previous notebook.

Here, we will change the default threshold to remove quasi-constant features, or, we should better say, features with low-variance:

Check the Scikit-learn docs for more details:
https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.VarianceThreshold.html

In [ ]:
#Tips: fit finds the features with low variance


VarianceThreshold(threshold=0.01)

In [ ]:
#Tips: get_support is a boolean vector that indicates which features 
# are retained, that is, which features have a higher variance than
# the threshold we indicated.

# Sum over get_support, to get the number
# of features that are not quasi-constant



215

In [ ]:
# Print the number of quasi-constant features



51

We can see that 51 columns / variables are almost constant. This means that 51 variables show predominantly one value for the majority of observations of the training set. Let's explore a few if these variables below.

In [ ]:
# Print the quasi constant variable names


Index(['var_1', 'var_2', 'var_7', 'var_9', 'var_10', 'var_19', 'var_28',
       'var_36', 'var_43', 'var_45', 'var_53', 'var_56', 'var_59', 'var_66',
       'var_67', 'var_69', 'var_71', 'var_104', 'var_106', 'var_116',
       'var_133', 'var_137', 'var_141', 'var_146', 'var_177', 'var_187',
       'var_189', 'var_194', 'var_197', 'var_198', 'var_202', 'var_218',
       'var_219', 'var_223', 'var_233', 'var_234', 'var_235', 'var_245',
       'var_247', 'var_249', 'var_250', 'var_251', 'var_256', 'var_260',
       'var_267', 'var_274', 'var_282', 'var_285', 'var_287', 'var_289',
       'var_298'],
      dtype='object')

In [ ]:
# crosscheck the percentage of observations showing each of the different values
# of the variable



0    0.999629
3    0.000200
6    0.000143
9    0.000029
Name: var_1, dtype: float64

We can see that > 99% of the observations show one value, 0. Therefore, this features is fairly constant.

In [ ]:
# Explore 'var_2' 

X_train['var_2'].value_counts() / np.float(len(X_train))

0    0.999971
1    0.000029
Name: var_2, dtype: float64



We can then remove the quasi-constant features utilizing the transform() method from the VarianceThreshold. Remember that this returns a NumPy array without feature names, so if we want a dataframe we need to reconstitute it.

In [ ]:
# capture feature names



In [ ]:
#remove the quasi-constant features



((35000, 215), (15000, 215))

By removing constant and almost constant features, we reduced the feature space from 300 to 215. This means, that 85 features were removed from this dataset. Almost a third!!

In [ ]:
# trasnform the array into a dataframe



var_3  var_4  var_5  var_6  var_8  var_11  var_12  var_13  var_14  var_15  \
0    0.0   2.79    0.0    0.0    0.0     0.0     0.0     0.0     0.0     3.0   
1    0.0   0.00    0.0    0.0    0.0     0.0     0.0     0.0     0.0     0.0   
2    0.0   2.94    0.0    0.0    0.0     0.0     0.0     0.0     0.0     3.0   
3    0.0   2.76    0.0    0.0    0.0     0.0     0.0     0.0     0.0     3.0   
4    0.0   2.94    0.0    0.0    0.0     0.0     0.0     0.0     0.0     3.0   

   ...  var_286  var_288  var_290  var_291  var_292  var_293  var_295  \
0  ...      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
1  ...      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
2  ...      0.0      0.0      0.0      0.0      0.0      3.0      0.0   
3  ...      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
4  ...      0.0      0.0      0.0      0.0      0.0      0.0      0.0   

   var_296  var_299  var_300  
0      0.0      0.0      0.0  
1      0.0      0.0      0.0  
2      0.0      0.0      0.0  
3      0.0      0.0      0.0  
4      0.0      0.0      0.0  

[5 rows x 215 columns]